WebGLRenderer: Add .renderShadowMap().

Oletus opened the pull request:
This can be used to manually update shadow maps, which can be convenient in case the user wants to set up the scene differently for shadow map rendering. This can be useful for example when showing a cutout view.

Fixed #23461

This contribution is funded by [Higharc](https://higharc.com/).

Oletus committed:
Add WebGLRenderer.renderShadowMap()

This can be used to manually update shadow maps, which can be convenient in case the user wants to set up the scene differently for shadow map rendering. This can be useful for example when showing a cutout view.

This work is sponsored by Higharc https://higharc.com/

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1053,6 +1053,37 @@ function WebGLRenderer( parameters = {} ) {
+
+		if ( _clippingEnabled === true ) clipping.endShadows( camera );
+
+		if ( renderStateStack.length > 0 ) {
```
It seems `renderStateStack.pop();` is missing right before this `if` statement.

Oletus replied:
Good catch, thanks!

Mugen87 commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1053,6 +1053,37 @@ function WebGLRenderer( parameters = {} ) {
 
 	};
 
+	this.renderShadowMap = function ( lightsWithShadows, scene, camera ) {
```
Since the method does not compute the world matrices of `scene` and `camera`, the application has to ensure proper up-to-date matrices, right?

Oletus replied:
True. I've added the auto-update code here. The application can decide whether it runs by setting autoUpdate as usual.

Mugen87 commented:
The example needs to be added to `files.json`. Otherwise it does not appear in the side menu.

And please add the respective E2E screenshot via:
```
npm run make-screenshot webgl_shadowmap_manual_update
```

Oletus commented:
@Mugen87 Added the file to the examples listing and added the screenshot, PR checks passed now!

WestLangley commented:
I had to diff the example in this PR to find the 3 lines which differ from the original example upon which it is based.

I think this will be confusing to users.

>This can be useful for example when showing a cutout view.

Why not create an example of that, instead?

RenaudRohlinger commented on src/renderers/WebGLRenderer.js:
```diff
@@ -1053,6 +1053,47 @@ function WebGLRenderer( parameters = {} ) {
+		if ( _clippingEnabled === true ) clipping.beginShadows();
+
+		shadowMap.needsUpdate = true;
+		shadowMap.render( lightsWithShadows, scene, camera, clipping );
```
clipping parameter does not exist in the ShadowMap render method

RenaudRohlinger commented:
I did give a try to this feature. It works like a charm. I was able to bake a static shadow on the whole scene in 4k and then render a small dynamic shadow in 1k that calls renderShadowMap only on a few meshes:

https://user-images.githubusercontent.com/15867665/159166252-c066c3dd-53a5-42e6-82ba-2e05135f3e5d.mov

Also, to make full usage of it I would recommend a 4th parameter `forceUpdate`

Like so:

`this.renderShadowMap = function ( lightsWithShadows, scene, camera, forceUpdate )`

That will allow being able to manually force to render an object even though `autoUpdate` and `needsUpdate` are set to false. We need to disable both these values if with use renderShadowMap otherwise it would clash with the natural state of how `shadowmap.render` works and call both methods per render.

In that case, the render method of `WebGLShadowMap` would have a third condition in order to prevent the render of the shadow:
`if ( shadow.autoUpdate === false && shadow.needsUpdate === false && forceUpdate === undefined ) continue;`

mrdoob commented:
> Why not create an example of that, instead?

Agreed. This needs a simple example.

RenaudRohlinger commented:
With that new feature it would be very interesting to also introduce Array usage on top of Mesh.Group for the second parameter of `WebGLShadowMap.render`.

It will allow to prevent a complex filtering of the meshes that we want to apply shadow to. That way, instead of doing `scene.traverse()`, we can just `renderShadowMap(lights, [meshA, meshB], camera, true)`

It could look like this:
```js

function WebGLShadowMap( _renderer, _objects, _capabilities ) {

	this.render = function ( lights, scene, camera, forceUpdate ) {
              
              // bypass if we want to manually update the shadowmap for specific elements
              if ( scope.autoUpdate === false && scope.needsUpdate === false && forceUpdate === undefined ) return;
              
              // allow arrays
              const elements = Array.isArray( scene ) ? scene : [ scene ];
             
              for ( let vp = 0; vp < viewportCount; vp ++ ) {
                    //viewport
                    _frustum = shadow.getFrustum();
              
                    for ( let j = 0; j < elements.length; j ++ ) {
                    
	                    const element = elements[ j ];
	                    renderObject( element, camera, shadow.camera, light, this.type );
                    
                    }
              
              }
              
              //....
```

Oletus committed:
Make a better example for shadow map manual updating

This example now includes one object that's in the shadow map, but isn't included in normal rendering. It's similar to a real use case where you'd disable a roof of a building when rendering an interior view, but still keep the whole building's shadow visible.

Oletus commented:
Rebased this to pass the tests and made sure that the patch still works, @mrdoob time to merge this? 😁 I'd be really happy if we got this change in, hope you're not too busy! 🙏

Oletus commented:
@mrdoob Wanted to check the status of this PR once again, it's been in the approved state for a long time! It seems like the code is still valid for the latest versions of three.js, but I could rebase the branch again if needed. We're still using this feature at Higharc, it would be great if we could get some of our customizations upstream.

mrdoob commented:
Sorry for the wait!

Would be great if you could rebase the PR yes.

Also, would it be possible to make it so `WebGLRenderer`'s `render()` also uses `renderShadowMap()`?

Samsy commented:
+1 !

Samsy commented:
Hey there, reviving this topic @RenaudRohlinger @sunag 
Could this still be work-able in the current state of the WebGPU backend ? 

I did hack my way on a custom threejs on the webGLRenderer to have both baked and real time lighting, this should definitely be useful to select which shadow you'd like to update, but also :

Determining if a light should contribute of the color lighting output, or cast shadow only

Example for a scene, let say you have two directional lights : 

1 used for baking shadows globally and for lighting color contribution

1 used for real time shadowing, but this one should not contribute to the color lighting as it would add up a second contribution, but on top affecting perfs since one light is only needed for the lighting color contribution

RenaudRohlinger commented:
Didn't try but I'm pretty sure if you render a shadow once and then do `light.shadow.autoUpdate = false` and `light.shadow.needsUpdate = false` on it, it should work just fine.
Unlike the WebGLRenderer, in the WebGPURenderer each `ShadowNode` that comes from a shadow is isolated and has its own rendering pipeline. @Samsy 

Also you probably need to use `"three": "mrdoob/three.js#dev",` in your dependency while waiting for a few fixes that will ship in r172:
https://github.com/mrdoob/three.js/pull/30081

Samsy commented:
@RenaudRohlinger Yes, It work this way indeed, I did edit the comment, I opened a feature request for the no-lightingcolor contribution

Samsy commented:
@RenaudRohlinger btw, about using "three": "mrdoob/three.js#dev", the builds are not updated
